micro-rdk/src/common/credentials_storage.rs

```diff
@@ -174,46 +195,82 @@ impl RAMStorage {
 
 #[cfg(feature = "ota")]
 impl OtaMetadataStorage for RAMStorage {
-    type Error = Infallible;
+    type Error = RAMStorageError;
     fn has_ota_metadata(&self) -> bool {
         let inner_ref = self.0.lock().unwrap();
         inner_ref.ota_metadata.is_some()
     }
-    fn store_ota_metadata(&self, ota_metadata: OtaMetadata) -> Result<(), Self::Error> {
+    fn store_ota_metadata(&self, ota_metadata: &OtaMetadata) -> Result<(), Self::Error> {
         let mut inner_ref = self.0.lock().unwrap();
-        let _ = inner_ref.ota_metadata.insert(ota_metadata);
+        let _ = inner_ref.ota_metadata.insert(ota_metadata.clone());
         Ok(())
     }
     fn get_ota_metadata(&self) -> Result<OtaMetadata, Self::Error> {
         let inner_ref = self.0.lock().unwrap();
-        Ok(inner_ref.ota_metadata.clone().unwrap_or_default())
+        inner_ref
+            .ota_metadata
+            .clone()
+            .ok_or(RAMStorageError::NotFound)
     }
     fn reset_ota_metadata(&self) -> Result<(), Self::Error> {
         let _ = self.0.lock().unwrap().ota_metadata.take();
         Ok(())
     }
 }
+#[cfg(feature = "ota")]
+impl<Iterable, Storage: OtaMetadataStorage> OtaMetadataStorage for Iterable
+where
+    for<'a> &'a Iterable: IntoIterator<Item = &'a Storage>,
+    Storage::Error: From<EmptyStorageCollectionError>,
+{
+    type Error = Storage::Error;
+    fn has_ota_metadata(&self) -> bool {
+        self.into_iter().any(OtaMetadataStorage::has_ota_metadata)
+    }
+    fn get_ota_metadata(&self) -> Result<OtaMetadata, Self::Error> {
+        self.into_iter().fold(
+            Err::<_, Self::Error>(EmptyStorageCollectionError.into()),
+            |val, s| val.or_else(|_| s.get_ota_metadata()),
+        )
+    }
+    fn reset_ota_metadata(&self) -> Result<(), Self::Error> {
+        self.into_iter().fold(
+            Err::<_, Self::Error>(EmptyStorageCollectionError.into()),
+            |val, s| val.or(s.reset_ota_metadata()),
```

Comment:
taking advantage of the eagerness of or might not be the best approach

Reply:
Can you elaborate a bit? I think I know what you are saying, but I'd rather not guess.

Reply:
As opposed to `or_else` `or` will always evaluate a function in argument if there is one (because it doesn't actually accept a closure) before considering replacing `val` hence the eagerness. We use to our advantage to make sure all copies of an object are erased through all storage. In theory we shouldn't need this since we use or_else elsewhere so only on copy should be in the storage array at anytime.